src/Misc/layoutroot/run.sh

```diff
@@ -11,54 +11,60 @@ fi
 # https://stackoverflow.com/questions/59895/getting-the-source-directory-of-a-bash-script-from-within
 SOURCE="${BASH_SOURCE[0]}"
 while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
-  DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
-  SOURCE="$(readlink "$SOURCE")"
-  [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
+    DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
+    SOURCE="$(readlink "$SOURCE")"
+    [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
 done
 DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
 
 # Do not "cd $DIR". For localRun, the current directory is expected to be the repo location on disk.
 
 # Run
 shopt -s nocasematch
-if [[ "$1" == "localRun" ]]; then
-    "$DIR"/bin/Runner.Listener $*
-else
-    "$DIR"/bin/Runner.Listener run $*
 
-# Return code 3 means the run once runner received an update message.
-# Sleep 5 seconds to wait for the update process finish
-    returnCode=$?
-    if [[ $returnCode == 3 ]]; then
-        if [ ! -x "$(command -v sleep)" ]; then
-            if [ ! -x "$(command -v ping)" ]; then
-                COUNT="0"
-                while [[ $COUNT != 5000 ]]; do
-                    echo "SLEEP" > /dev/null
-                    COUNT=$[$COUNT+1]
-                done
-            else
-                ping -c 5 127.0.0.1 > /dev/null
-            fi
-        else
-            sleep 5
-        fi
-    elif [[ $returnCode == 4 ]]; then
-        if [ ! -x "$(command -v sleep)" ]; then
-            if [ ! -x "$(command -v ping)" ]; then
-                COUNT="0"
-                while [[ $COUNT != 5000 ]]; do
-                    echo "SLEEP" > /dev/null
-                    COUNT=$[$COUNT+1]
-                done
-            else
-                ping -c 5 127.0.0.1 > /dev/null
-            fi
+safe_sleep() {
+    if [ ! -x "$(command -v sleep)" ]; then
+        if [ ! -x "$(command -v ping)" ]; then
+            COUNT="0"
+            while [[ $COUNT != 5000 ]]; do
+                echo "SLEEP" > /dev/null
+                COUNT=$[$COUNT+1]
+            done
         else
-            sleep 5
+            ping -c 5 127.0.0.1 > /dev/null
         fi
+    else
+        sleep 5
+    fi
+}
+
+while :;
+do
+    if [[ "$1" == "localRun" ]]; then
+        "$DIR"/bin/Runner.Listener $*
+    else
         "$DIR"/bin/Runner.Listener run $*
+    fi
+    
+    returnCode=$?
+    if [[ $returnCode == 0 ]]; then
+        echo "Runner listener exit with 0 return code, stop the service, no retry needed."
+        exit $returnCode
+    elif [[ $returnCode == 1 ]]; then
+        echo "Runner listener exit with terminated error, stop the service, no retry needed."
+        exit $returnCode
+    elif [[ $returnCode == 2 ]]; then
```

Comment:
Now we would retry on `$returnCode == 2` in run.sh|cmd as well, just like in RunnerService.js